framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,152 @@
+        if (tree == null) {
+            return;
+        }
+        AnnotatedTypeMirror oldType = typeFactory.getAnnotatedType(tree);
+        AnnotationMirror newAnno = getCombinedAnno(oldType, Arrays.asList(values));
+        insertIntoStores(result, node, newAnno);
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+            MethodInvocationNode methodInvocationNode = (MethodInvocationNode) node;
+            Node receiver = methodInvocationNode.getTarget().getReceiver();
+            MethodInvocationTree invokedMethod = (MethodInvocationTree) tree;
+
+            while (receiver != null && typeFactory.returnsThis(invokedMethod)) {
+                // Note that this call doesn't do anything if receiver is a method call
+                // that is not deterministic, though it can still continue to recurse.
+                insertIntoStores(result, receiver, newAnno);
+
+                Tree receiverTree = receiver.getTree();
+                // Possibly recurse: if the receiver is itself a method call,
+                // then we need to also propagate this new information to its receiver
+                // if the method being called has an @This return type.
+                //
+                // Note that we must check for null, because the tree could be
+                // implicit (when calling an instance method on the class itself).
+                // In that case, do not attempt to refine either - the receiver is
+                // not a method invocation, anyway.
+                if (receiverTree == null || receiverTree.getKind() != Tree.Kind.METHOD_INVOCATION) {
+                    // Do not continue, because the receiver isn't a method invocation itself. The
+                    // end of the chain of calls has been reached.
+                    break;
+                }
+
+                invokedMethod = (MethodInvocationTree) receiver.getTree();
+                receiver = ((MethodInvocationNode) receiver).getTarget().getReceiver();
+            }
+        }
+    }
+
+    /**
+     * Unions the values in oldType with the values in newValues to produce a single accumulator
+     * type qualifier.
+     *
+     * @param oldType an annotated type mirror whose values should be included, or null which is
+     *     treated as top
+     * @param newValues new values to include
+     * @return an annotation representing all the values
+     */
+    private AnnotationMirror getCombinedAnno(
+            @Nullable AnnotatedTypeMirror oldType, List<String> newValues) {
+        AnnotationMirror oldAnno;
+        if (oldType == null) {
+            oldAnno = typeFactory.top;
+        } else {
+            oldAnno = oldType.getAnnotationInHierarchy(typeFactory.top);
+            if (oldAnno == null) {
+                oldAnno = typeFactory.top;
+            }
+        }
+        List<String> allValues = new ArrayList<>(newValues);
+        if (typeFactory.isAccumulatorAnnotation(oldAnno)) {
```

Comment:
I think this is a way of saying "not top or bottom".  It's surprising that combining any A with top has the same effect as combining A with bottom.  Is that intentional?  Should combining A with bottom lead to bottom?

Reply:
I rewrote this method to make this test unnecessary (it now uses GLB).